File: SkillHub/Catalog/Services/UsageLog.swift
import Foundation

/// Counts agent invocations by reading Cursor, Claude, and Codex session transcripts.
/// Cursor/Codex use a skill by reading `SKILL.md`; Claude also has a dedicated `Skill` tool.
nonisolated enum UsageLog {
    /// Compiled once; Regex is not Sendable but these values are immutable after init.
    nonisolated(unsafe) private static let skillPath = try! Regex(#"(?:skills|skills-cursor|default-skills)/([A-Za-z0-9][A-Za-z0-9._-]{0,80})/SKILL\.md"#)
    nonisolated(unsafe) private static let promptPath = try! Regex(#"(?:\.codex/prompts|\.skill-hub/library/prompts)/([A-Za-z0-9][A-Za-z0-9._-]{0,80})\.md"#)

    private static let readTools: Set<String> = ["read", "read_file", "readfile", "view"]
    private static let skillTools: Set<String> = ["skill"]
    private static let skipTools: Set<String> = [
        "grep", "glob", "search", "write", "strreplace", "edit", "delete",
        "todowrite", "websearch", "webfetch"
    ]

    static func loadIndex(cacheURL: URL) -> UsageIndex {
        merge(loadCache(cacheURL))
    }

    /// Reuses per-file cache entries when mtime and size are unchanged. First run is slow; later ones are a directory walk.
    static func scan(home: URL, cacheURL: URL) -> UsageIndex {
        let previous = loadCache(cacheURL)
        var next: [String: UsageFileRecord] = [:]
        next.reserveCapacity(previous.count)
        var dirty = false
        let fm = FileManager.default
        for root in roots(home: home) where fm.fileExists(atPath: root.path) {
            guard let enumerator = fm.enumerator(
                at: root,
                includingPropertiesForKeys: [.isRegularFileKey, .contentModificationDateKey, .fileSizeKey],
                options: [.skipsHiddenFiles]
            ) else { continue }
            while let url = enumerator.nextObject() as? URL {
                guard url.pathExtension == "jsonl" else { continue }
                let path = url.path
                if path.contains("/.cursor/") && !path.contains("/agent-transcripts/") { continue }
                let values = try? url.resourceValues(forKeys: [.contentModificationDateKey, .fileSizeKey])
                let mtime = values?.contentModificationDate?.timeIntervalSince1970 ?? 0
                let size = values?.fileSize ?? 0
                let key = url.standardizedFileURL.path
                if let cached = previous[key], abs(cached.modifiedAt - mtime) < 1, cached.size == size {
                    next[key] = cached
                    continue
                }
                var piece = UsageIndex()
                ingest(file: url, fileDate: Date(timeIntervalSince1970: mtime), into: &piece)
                next[key] = UsageFileRecord(modifiedAt: mtime, size: size, skills: piece.skills, prompts: piece.prompts)
                dirty = true
            }
        }
        if next.count != previous.count { dirty = true }
        if dirty { saveCache(next, to: cacheURL) }
        return merge(next)
    }

    private static func roots(home: URL) -> [URL] {
        [
            home.appending(path: ".cursor/projects", directoryHint: .isDirectory),
            home.appending(path: ".claude/projects", directoryHint: .isDirectory),
            home.appending(path: ".codex/sessions", directoryHint: .isDirectory)
        ]
    }

    private static func merge(_ files: [String: UsageFileRecord]) -> UsageIndex {
        var index = UsageIndex()
        for record in files.values {
            var piece = UsageIndex()
            piece.skills = record.skills
            piece.prompts = record.prompts
            index.merge(piece)
        }
        return index
    }

    private static func loadCache(_ url: URL) -> [String: UsageFileRecord] {
        guard let data = try? Data(contentsOf: url) else { return [:] }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return (try? decoder.decode(UsageFileCache.self, from: data))?.files ?? [:]
    }

    private static func saveCache(_ files: [String: UsageFileRecord], to url: URL) {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        guard let data = try? encoder.encode(UsageFileCache(files: files)) else { return }
        try? FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        try? data.write(to: url, options: .atomic)
    }

    static func ingest(file url: URL, fileDate: Date, into index: inout UsageIndex) {
        guard let text = try? String(contentsOf: url, encoding: .utf8) else { return }
        for line in text.split(whereSeparator: \.isNewline) {
            ingest(line: String(line), fileDate: fileDate, into: &index)
        }
    }

    static func ingest(line: String, fileDate: Date, into index: inout UsageIndex) {
        if !line.contains("SKILL.md")
            && !line.contains("\"Skill\"")
            && !line.contains("library/prompts")
            && !line.contains(".codex/prompts")
        {
            return
        }
        guard let data = line.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data)
        else {
            recordPaths(in: line, at: fileDate, into: &index)
            return
        }
        walk(object, at: date(in: object) ?? fileDate, into: &index)
    }

    private static func walk(_ value: Any, at date: Date, into index: inout UsageIndex) {
        if let dictionary = value as? [String: Any] {
            if consider(dictionary, at: date, into: &index) { return }
            for nested in dictionary.values { walk(nested, at: date, into: &index) }
        } else if let array = value as? [Any] {
            for nested in array { walk(nested, at: date, into: &index) }
        }
    }

    /// Returns true when this object is a tool call we already handled (do not walk children).
    private static func consider(_ dictionary: [String: Any], at date: Date, into index: inout UsageIndex) -> Bool {
        let type = (dictionary["type"] as? String)?.lowercased()
        if type == "tool_result" || type == "function_call_output" { return true }

        let isCall = type == "tool_use" || type == "function_call" || type == "custom_tool_call"
            || dictionary["name"] != nil && dictionary["input"] != nil
        guard isCall, let name = dictionary["name"] as? String else { return false }

        let folded = name.lowercased()
        if skipTools.contains(folded) { return true }

        let input = dictionary["input"] ?? dictionary["arguments"]
        if skillTools.contains(folded) {
            if let skill = skillName(fromInput: input) {
                index.addSkill(skill, at: date)
            }
            return true
        }
        if readTools.contains(folded) {
            if let path = path(fromInput: input) {
                recordPaths(in: path, at: date, into: &index)
            }
            return true
        }
        if folded == "exec" || folded == "exec_command" || folded == "bash" || folded == "shell" {
            if let text = string(from: input) {
                recordPaths(in: text, at: date, into: &index)
            }
            return true
        }
        return false
    }

    private static func skillName(fromInput input: Any?) -> String? {
        if let dictionary = input as? [String: Any] {
            for key in ["skill", "name", "skill_name", "skillName"] {
                if let value = dictionary[key] as? String, !value.isEmpty { return value }
            }
        }
        return nil
    }

    private static func path(fromInput input: Any?) -> String? {
        if let dictionary = input as? [String: Any] {
            for key in ["path", "file_path", "filePath", "target_file", "targetFile"] {
                if let value = dictionary[key] as? String, !value.isEmpty { return value }
            }
        }
        return string(from: input)
    }

    private static func string(from input: Any?) -> String? {
        switch input {
        case let text as String: text
        case let dictionary as [String: Any]:
            (dictionary["command"] as? String)
                ?? (dictionary["cmd"] as? String)
                ?? (dictionary["input"] as? String)
        default: nil
        }
    }

    static func recordPaths(in text: String, at date: Date, into index: inout UsageIndex) {
        for match in text.matches(of: skillPath) {
            if let name = match.output[1].substring { index.addSkill(String(name), at: date) }
        }
        for match in text.matches(of: promptPath) {
            if let name = match.output[1].substring { index.addPrompt(String(name), at: date) }
        }
    }

    private static func date(in value: Any) -> Date? {
        guard let dictionary = value as? [String: Any] else { return nil }
        if let stamp = dictionary["timestamp"] as? String { return parseDate(stamp) }
        if let nested = dictionary["payload"] { return date(in: nested) }
        if let nested = dictionary["message"] { return date(in: nested) }
        return nil
    }

    private static func parseDate(_ raw: String) -> Date? {
        if let date = try? Date.ISO8601FormatStyle(includingFractionalSeconds: true).parse(raw) { return date }
        return try? Date.ISO8601FormatStyle().parse(raw)
    }
}

nonisolated struct UsageFileRecord: Codable, Sendable, Equatable {
    var modifiedAt: TimeInterval
    var size: Int
    var skills: [String: UsageStat]
    var prompts: [String: UsageStat]
}

nonisolated struct UsageFileCache: Codable, Sendable {
    var files: [String: UsageFileRecord]
}

File: windows/src/SkillHub.Core/UsageLog.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Text.RegularExpressions;

namespace SkillHub.Core;

public sealed class UsageStat
{
    public int Count { get; set; }
    public DateTime? LastInvokedAt { get; set; }

    public static UsageStat Empty => new();

    public void Add(DateTime? date)
    {
        Count++;
        if (date is { } when && (LastInvokedAt == null || when > LastInvokedAt))
        {
            LastInvokedAt = when;
        }
    }
}

public sealed class UsageIndex
{
    public Dictionary<string, UsageStat> Skills { get; } = new(StringComparer.OrdinalIgnoreCase);
    public Dictionary<string, UsageStat> Prompts { get; } = new(StringComparer.OrdinalIgnoreCase);

    public static UsageIndex Empty { get; } = new();

    public UsageStat Skill(SkillItem item) =>
        Lookup(Skills, item.FolderName) ?? Lookup(Skills, item.Name) ?? UsageStat.Empty;

    public UsageStat Prompt(PromptItem item)
    {
        var stem = Path.GetFileNameWithoutExtension(item.CanonicalPath);
        return Lookup(Prompts, stem) ?? Lookup(Prompts, item.Title) ?? UsageStat.Empty;
    }

    public void AddSkill(string raw, DateTime? date)
    {
        var name = raw.Trim();
        if (name.Length == 0)
        {
            return;
        }

        Add(Skills, name, date);
        var colon = name.LastIndexOf(':');
        if (colon >= 0 && colon < name.Length - 1)
        {
            Add(Skills, name[(colon + 1)..], date);
        }
    }

    public void AddPrompt(string raw, DateTime? date) => Add(Prompts, raw, date);

    public void Merge(UsageIndex other)
    {
        MergeInto(Skills, other.Skills);
        MergeInto(Prompts, other.Prompts);
    }

    static void MergeInto(Dictionary<string, UsageStat> dest, Dictionary<string, UsageStat> src)
    {
        foreach (var (key, stat) in src)
        {
            if (!dest.TryGetValue(key, out var current))
            {
                dest[key] = new UsageStat { Count = stat.Count, LastInvokedAt = stat.LastInvokedAt };
                continue;
            }

            current.Count += stat.Count;
            if (stat.LastInvokedAt is { } at && (current.LastInvokedAt == null || at > current.LastInvokedAt))
            {
                current.LastInvokedAt = at;
            }
        }
    }

    static UsageStat? Lookup(Dictionary<string, UsageStat> map, string raw) =>
        map.TryGetValue(raw.Trim(), out var stat) ? stat : null;

    static void Add(Dictionary<string, UsageStat> map, string raw, DateTime? date)
    {
        var key = raw.Trim();
        if (key.Length == 0)
        {
            return;
        }

        if (!map.TryGetValue(key, out var stat))
        {
            stat = new UsageStat();
            map[key] = stat;
        }

        stat.Add(date);
    }
}

public static class UsageLog
{
    static readonly Regex SkillPath = new(
        @"(?:skills|skills-cursor|default-skills)/([A-Za-z0-9][A-Za-z0-9._-]{0,80})/SKILL\.md",
        RegexOptions.Compiled | RegexOptions.CultureInvariant);
    static readonly Regex PromptPath = new(
        @"(?:\.codex/prompts|\.skill-hub/library/prompts)/([A-Za-z0-9][A-Za-z0-9._-]{0,80})\.md",
        RegexOptions.Compiled | RegexOptions.CultureInvariant | RegexOptions.IgnoreCase);

    static readonly HashSet<string> ReadTools = new(StringComparer.OrdinalIgnoreCase)
    {
        "read", "read_file", "readfile", "view"
    };
    static readonly HashSet<string> SkillTools = new(StringComparer.OrdinalIgnoreCase) { "skill" };
    static readonly HashSet<string> SkipTools = new(StringComparer.OrdinalIgnoreCase)
    {
        "grep", "glob", "search", "write", "strreplace", "edit", "delete",
        "todowrite", "websearch", "webfetch"
    };

    static readonly JsonSerializerOptions CacheOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
    };

    public static UsageIndex LoadIndex(string cachePath) => Merge(LoadCache(cachePath));

    public static UsageIndex Scan(string home, string cachePath)
    {
        var previous = LoadCache(cachePath);
        var next = new Dictionary<string, UsageFileRecord>(StringComparer.Ordinal);
        var dirty = false;
        foreach (var file in EnumerateLogs(home))
        {
            FileInfo info;
            try
            {
                info = new FileInfo(file);
            }
            catch (Exception)
            {
                continue;
            }

            if (!info.Exists)
            {
                continue;
            }

            var key = info.FullName;
            if (previous.TryGetValue(key, out var cached)
                && cached.Size == info.Length
                && Math.Abs((cached.ModifiedAt - info.LastWriteTimeUtc).TotalSeconds) < 1)
            {
                next[key] = cached;
                continue;
            }

            var piece = new UsageIndex();
            IngestFile(file, piece);
            next[key] = new UsageFileRecord
            {
                ModifiedAt = info.LastWriteTimeUtc,
                Size = info.Length,
                Skills = new Dictionary<string, UsageStat>(piece.Skills, StringComparer.OrdinalIgnoreCase),
                Prompts = new Dictionary<string, UsageStat>(piece.Prompts, StringComparer.OrdinalIgnoreCase)
            };
            dirty = true;
        }

        if (next.Count != previous.Count)
        {
            dirty = true;
        }

        if (dirty)
        {
            SaveCache(cachePath, next);
        }

        return Merge(next);
    }

    static IEnumerable<string> EnumerateLogs(string home)
    {
        foreach (var root in new[]
                 {
                     Path.Combine(home, ".cursor", "projects"),
                     Path.Combine(home, ".claude", "projects"),
                     Path.Combine(home, ".codex", "sessions")
                 })
        {
            if (!Directory.Exists(root))
            {
                continue;
            }

            IEnumerable<string> files;
            try
            {
                files = Directory.EnumerateFiles(root, "*.jsonl", SearchOption.AllDirectories);
            }
            catch (Exception)
            {
                continue;
            }

            foreach (var file in files)
            {
                var normalized = file.Replace('\\', '/');
                if (normalized.Contains("/.cursor/", StringComparison.OrdinalIgnoreCase)
                    && !normalized.Contains("/agent-transcripts/", StringComparison.OrdinalIgnoreCase))
                {
                    continue;
                }

                yield return file;
            }
        }
    }

    static UsageIndex Merge(Dictionary<string, UsageFileRecord> files)
    {
        var index = new UsageIndex();
        foreach (var record in files.Values)
        {
            var piece = new UsageIndex();
            foreach (var (key, stat) in record.Skills)
            {
                piece.Skills[key] = stat;
            }

            foreach (var (key, stat) in record.Prompts)
            {
                piece.Prompts[key] = stat;
            }

            index.Merge(piece);
        }

        return index;
    }

    static Dictionary<string, UsageFileRecord> LoadCache(string path)
    {
        try
        {
            if (!File.Exists(path))
            {
                return new Dictionary<string, UsageFileRecord>(StringComparer.Ordinal);
            }

            var json = File.ReadAllText(path);
            return JsonSerializer.Deserialize<UsageFileCache>(json, CacheOptions)?.Files
                   ?? new Dictionary<string, UsageFileRecord>(StringComparer.Ordinal);
        }
        catch (Exception)
        {
            return new Dictionary<string, UsageFileRecord>(StringComparer.Ordinal);
        }
    }

    static void SaveCache(string path, Dictionary<string, UsageFileRecord> files)
    {
        try
        {
            Directory.CreateDirectory(Path.GetDirectoryName(path)!);
            var json = JsonSerializer.Serialize(new UsageFileCache { Files = files }, CacheOptions);
            PathUtil.WriteAtomic(path, json);
        }
        catch (Exception)
        {
            // Cache is optional; the next scan can rebuild it.
        }
    }

    public static void IngestLine(string line, DateTime fileDate, UsageIndex index)
    {
        if (!line.Contains("SKILL.md", StringComparison.Ordinal)
            && !line.Contains("\"Skill\"", StringComparison.Ordinal)
            && !line.Contains("library/prompts", StringComparison.Ordinal)
            && !line.Contains(".codex/prompts", StringComparison.Ordinal))
        {
            return;
        }

        try
        {
            using var doc = JsonDocument.Parse(line);
            Walk(doc.RootElement, DateOf(doc.RootElement) ?? fileDate, index);
        }
        catch (JsonException)
        {
            RecordPaths(line, fileDate, index);
        }
    }

    static void IngestFile(string path, UsageIndex index)
    {
        DateTime stamp;
        try
        {
            stamp = File.GetLastWriteTime(path);
        }
        catch (Exception)
        {
            return;
        }

        IEnumerable<string> lines;
        try
        {
            lines = File.ReadLines(path);
        }
        catch (Exception)
        {
            return;
        }

        foreach (var line in lines)
        {
            IngestLine(line, stamp, index);
        }
    }

    static void Walk(JsonElement value, DateTime date, UsageIndex index)
    {
        switch (value.ValueKind)
        {
            case JsonValueKind.Object:
                if (Consider(value, date, index))
                {
                    return;
                }

                foreach (var property in value.EnumerateObject())
                {
                    Walk(property.Value, date, index);
                }

                break;
            case JsonValueKind.Array:
                foreach (var item in value.EnumerateArray())
                {
                    Walk(item, date, index);
                }

                break;
        }
    }

    static bool Consider(JsonElement dictionary, DateTime date, UsageIndex index)
    {
        var type = dictionary.TryGetProperty("type", out var typeEl) && typeEl.ValueKind == JsonValueKind.String
            ? typeEl.GetString()
            : null;
        if (type is "tool_result" or "function_call_output")
        {
            return true;
        }

        var hasName = dictionary.TryGetProperty("name", out var nameEl) && nameEl.ValueKind == JsonValueKind.String;
        var hasInput = dictionary.TryGetProperty("input", out _) || dictionary.TryGetProperty("arguments", out _);
        var isCall = type is "tool_use" or "function_call" or "custom_tool_call" || hasName && hasInput;
        if (!isCall || !hasName)
        {
            return false;
        }

        var name = nameEl.GetString() ?? "";
        if (SkipTools.Contains(name))
        {
            return true;
        }

        JsonElement input = default;
        var gotInput = dictionary.TryGetProperty("input", out input) || dictionary.TryGetProperty("arguments", out input);

        if (SkillTools.Contains(name))
        {
            if (gotInput && SkillName(input) is { } skill)
            {
                index.AddSkill(skill, date);
            }

            return true;
        }

        if (ReadTools.Contains(name))
        {
            if (gotInput && PathFrom(input) is { } path)
            {
                RecordPaths(path, date, index);
            }

            return true;
        }

        if (name.Equals("exec", StringComparison.OrdinalIgnoreCase)
            || name.Equals("exec_command", StringComparison.OrdinalIgnoreCase)
            || name.Equals("bash", StringComparison.OrdinalIgnoreCase)
            || name.Equals("shell", StringComparison.OrdinalIgnoreCase))
        {
            if (gotInput && StringFrom(input) is { } text)
            {
                RecordPaths(text, date, index);
            }

            return true;
        }

        return false;
    }

    static string? SkillName(JsonElement input)
    {
        if (input.ValueKind != JsonValueKind.Object)
        {
            return null;
        }

        foreach (var key in new[] { "skill", "name", "skill_name", "skillName" })
        {
            if (input.TryGetProperty(key, out var value) && value.ValueKind == JsonValueKind.String)
            {
                var text = value.GetString();
                if (!string.IsNullOrWhiteSpace(text))
                {
                    return text;
                }
            }
        }

        return null;
    }

    static string? PathFrom(JsonElement input)
    {
        if (input.ValueKind == JsonValueKind.Object)
        {
            foreach (var key in new[] { "path", "file_path", "filePath", "target_file", "targetFile" })
            {
                if (input.TryGetProperty(key, out var value) && value.ValueKind == JsonValueKind.String)
                {
                    var text = value.GetString();
                    if (!string.IsNullOrWhiteSpace(text))
                    {
                        return text;
                    }
                }
            }
        }

        return StringFrom(input);
    }

    static string? StringFrom(JsonElement input) => input.ValueKind switch
    {
        JsonValueKind.String => input.GetString(),
        JsonValueKind.Object when input.TryGetProperty("command", out var command) => command.GetString(),
        JsonValueKind.Object when input.TryGetProperty("cmd", out var cmd) => cmd.GetString(),
        JsonValueKind.Object when input.TryGetProperty("input", out var nested) => nested.GetString(),
        _ => null
    };

    public static void RecordPaths(string text, DateTime date, UsageIndex index)
    {
        foreach (Match match in SkillPath.Matches(text))
        {
            index.AddSkill(match.Groups[1].Value, date);
        }

        foreach (Match match in PromptPath.Matches(text))
        {
            index.AddPrompt(match.Groups[1].Value, date);
        }
    }

    static DateTime? DateOf(JsonElement value)
    {
        if (value.ValueKind != JsonValueKind.Object)
        {
            return null;
        }

        if (value.TryGetProperty("timestamp", out var stamp) && stamp.ValueKind == JsonValueKind.String
            && DateTime.TryParse(stamp.GetString(), out var parsed))
        {
            return parsed.ToLocalTime();
        }

        if (value.TryGetProperty("payload", out var payload))
        {
            return DateOf(payload);
        }

        if (value.TryGetProperty("message", out var message))
        {
            return DateOf(message);
        }

        return null;
    }
}

sealed class UsageFileCache
{
    public Dictionary<string, UsageFileRecord> Files { get; set; } = new(StringComparer.Ordinal);
}

sealed class UsageFileRecord
{
    public DateTime ModifiedAt { get; set; }
    public long Size { get; set; }
    public Dictionary<string, UsageStat> Skills { get; set; } = new(StringComparer.OrdinalIgnoreCase);
    public Dictionary<string, UsageStat> Prompts { get; set; } = new(StringComparer.OrdinalIgnoreCase);
}
